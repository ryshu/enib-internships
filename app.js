"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const express_1 = __importDefault(require("express"));
const body_parser_1 = __importDefault(require("body-parser"));
const compression_1 = __importDefault(require("compression")); // compresses requests
const lusca_1 = __importDefault(require("lusca"));
const dotenv_1 = __importDefault(require("dotenv"));
const express_flash_1 = __importDefault(require("express-flash"));
const path_1 = __importDefault(require("path"));
const method_override_1 = __importDefault(require("method-override"));
// Load environment variables from .env file, where API keys and passwords are configured
dotenv_1.default.config({ path: '.env' });
// Loggers
const logger_1 = __importDefault(require("./utils/logger"));
const error_1 = require("./utils/error");
const http_status_codes_1 = __importDefault(require("http-status-codes"));
// router and database
const route_1 = __importDefault(require("./configs/setup/route"));
const cas_1 = __importDefault(require("./configs/setup/cas"));
require("./configs/instances/database"); // Only import to setup
require("./configs/setup/database"); // Only import to setup
const express_session_1 = __importDefault(require("express-session"));
const handle_1 = require("./auth/cas/handle");
// Create Express server
const app = express_1.default();
exports.session = express_session_1.default({
    secret: process.env.SESSION_SECRET,
    resave: false,
    saveUninitialized: true,
});
// Express configuration
app.set('port', process.env.INTERNSHIP_ENIB_API_PORT || 3000);
app.use(compression_1.default());
app.use(body_parser_1.default.json({ limit: '1mb' }));
app.use(body_parser_1.default.urlencoded({ extended: true, limit: '1mb' }));
app.use(express_flash_1.default());
app.use(lusca_1.default.xframe('SAMEORIGIN'));
app.use(lusca_1.default.xssProtection(true));
app.use(exports.session);
// CAS Setup
app.get('/logout', cas_1.default.logout); // Logout pass
app.use('/', cas_1.default.bounce, // CAS bounce to connect user if isn't
(req, _res, next) => {
    if (!req.session.info) {
        // If no session info, handle connection
        handle_1.handleConnection(req)
            .then(() => next())
            .catch((e) => next(e));
    }
    else {
        // else, pass
        next();
    }
}, express_1.default.static(path_1.default.join(__dirname, 'public'), { maxAge: 0 }));
app.use(route_1.default);
/**
 * Error handler setup
 */
app.use(method_override_1.default());
app.use((err, _req, res, _next) => {
    if (err instanceof error_1.APIError) {
        return res.status(err.status).json(err);
    }
    else {
        logger_1.default.error(err);
        return res.status(http_status_codes_1.default.INTERNAL_SERVER_ERROR).send(`Internal server error`);
    }
});
exports.default = app;
//# sourceMappingURL=app.js.map