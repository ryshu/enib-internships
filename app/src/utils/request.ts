import axios from 'axios';
import { Message, MessageBox } from 'element-ui';

const service = axios.create({
  baseURL: process.env.VUE_APP_BASE_API, // url = base url + request url
  timeout: 5000,
  // withCredentials: true // send cookies when cross-domain requests
});

// Request interceptors
service.interceptors.request.use(
  config => config,
  error => {
    Promise.reject(error);
  }
);

// Response interceptors
service.interceptors.response.use(
  response => {
    const res = response.data;
    if (res.code && res.code !== 200 && res.code !== 201) {
      Message({
        message: res.message || 'Error',
        type: 'error',
        duration: 5 * 1000,
      });
      if (res.code === 500) {
        MessageBox.confirm(
          '你已被登出，可以取消继续留在该页面，或者重新登录',
          '确定登出',
          {
            confirmButtonText: '重新登录',
            cancelButtonText: '取消',
            type: 'warning',
          }
        ).then(() => {
          window.location.assign('/logout');
        });
      }
      return Promise.reject(new Error(res.message || 'Error'));
    } else {
      return response.data;
    }
  },
  error => {
    console.log(error);
    Message({
      message: error.message,
      type: 'error',
      duration: 5 * 1000,
    });
    return Promise.reject(error);
  }
);

export default service;
