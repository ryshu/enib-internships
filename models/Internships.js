"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Sequelize = __importStar(require("sequelize"));
const database_1 = __importDefault(require("../configs/instances/database"));
class Internships extends Sequelize.Model {
}
Internships.init({
    id: {
        type: Sequelize.DataTypes.INTEGER.UNSIGNED,
        autoIncrement: true,
        primaryKey: true,
    },
    // Data
    subject: {
        type: new Sequelize.DataTypes.STRING(256),
        allowNull: false,
    },
    description: {
        type: new Sequelize.DataTypes.TEXT(),
        allowNull: true,
    },
    // Localisation
    country: {
        type: new Sequelize.DataTypes.STRING(128),
        allowNull: false,
    },
    city: {
        type: new Sequelize.DataTypes.STRING(128),
        allowNull: false,
    },
    postalCode: {
        type: new Sequelize.DataTypes.STRING(128),
        allowNull: false,
    },
    address: {
        type: new Sequelize.DataTypes.STRING(128),
        allowNull: false,
    },
    additional: {
        type: new Sequelize.DataTypes.STRING(),
        allowNull: true,
    },
    // State
    isInternshipAbroad: {
        type: Sequelize.DataTypes.BOOLEAN,
        allowNull: false,
        defaultValue: false,
    },
    isValidated: {
        type: Sequelize.DataTypes.BOOLEAN,
        allowNull: false,
        defaultValue: false,
    },
    // Date
    startAt: {
        type: Sequelize.DataTypes.INTEGER,
        allowNull: true,
        defaultValue: null,
    },
    endAt: {
        type: Sequelize.DataTypes.INTEGER,
        allowNull: true,
        defaultValue: null,
    },
}, {
    tableName: 'internships',
    sequelize: database_1.default,
    defaultScope: {
        attributes: {
            exclude: [
                'businessId',
                'studentId',
                'categoryId',
                'availableCampaignId',
                'validatedCampaignId',
                'mentorId',
            ],
        },
    },
});
exports.default = Internships;
//# sourceMappingURL=Internships.js.map