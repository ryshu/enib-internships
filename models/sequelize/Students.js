"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Sequelize = __importStar(require("sequelize"));
const database_1 = __importDefault(require("../../configs/instances/database"));
class Students extends Sequelize.Model {
}
Students.init({
    id: {
        type: Sequelize.DataTypes.INTEGER.UNSIGNED,
        autoIncrement: true,
        primaryKey: true,
    },
    firstName: {
        type: new Sequelize.DataTypes.STRING(128),
        allowNull: false,
    },
    lastName: {
        type: new Sequelize.DataTypes.STRING(128),
        allowNull: false,
    },
    email: {
        type: new Sequelize.DataTypes.STRING(128),
        allowNull: false,
    },
    semester: {
        type: new Sequelize.DataTypes.STRING(3),
        allowNull: false,
    },
}, {
    tableName: 'students',
    sequelize: database_1.default,
    timestamps: true,
    paranoid: true,
});
exports.default = Students;
//# sourceMappingURL=Students.js.map