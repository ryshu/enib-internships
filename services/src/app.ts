import express, { Request, Response, NextFunction } from 'express';
import bodyParser from 'body-parser';
import cors from 'cors';
import compression from 'compression'; // compresses requests
import lusca from 'lusca';
import dotenv from 'dotenv';
import flash from 'express-flash';
import path from 'path';
import methodOverride from 'method-override';

// Load environment variables from .env file, where API keys and passwords are configured
dotenv.config({ path: '.env' });

// Loggers
import logger from './utils/logger';
import { APIError } from './utils/error';
import httpStatus from 'http-status-codes';

// router and database
import router from './configs/setup/route';
import './configs/instances/database'; // Only import to setup
import './configs/setup/database'; // Only import to setup

// Create Express server
const app = express();

// Express configuration
app.set('port', process.env.INTERNSHIP_ENIB_API_PORT || 3000);
app.use(compression());
app.use(bodyParser.json({ limit: '1mb' }));
app.use(bodyParser.urlencoded({ extended: true, limit: '1mb' }));
app.use(flash());
app.use(lusca.xframe('SAMEORIGIN'));
app.use(lusca.xssProtection(true));

app.use(express.static(path.join(__dirname, 'public'), { maxAge: 31557600000 }));

const corsOptions = {
    origin: 'http://localhost:4200',
    optionsSuccessStatus: 200,
};
app.use(cors(corsOptions));

app.use(router);

/**
 * Error handler setup
 */
app.use(methodOverride());
app.use((err: APIError | Error, _req: Request, res: Response, _next: NextFunction) => {
    if (err instanceof APIError) {
        return res.status(err.status).json(err);
    } else {
        logger.error(err);
        return res.status(httpStatus.INTERNAL_SERVER_ERROR).send(`Internal server error`);
    }
});

export default app;
