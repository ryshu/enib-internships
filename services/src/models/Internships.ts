import * as Sequelize from 'sequelize';

import database from '../configs/instances/database';

import Businesses from './Businesses';
import InternshipTypes from './InternshipTypes';

class Internships extends Sequelize.Model implements IInternshipEntity {
    public static associations: {
        category: Sequelize.Association<Internships, InternshipTypes>;
        business: Sequelize.Association<Internships, Businesses>;
    };

    public id!: number; // Note that the `null assertion` `!` is required in strict mode.

    // Data
    public subject: string;
    public description: string;

    // Localisation
    public country: string;
    public city: string;
    public postalCode: string;
    public address: string;
    public additional?: string;

    // State
    public isInternshipAbroad: boolean;
    public isValidated: boolean;

    // timestamps!
    public readonly createdAt!: Date;
    public readonly updatedAt!: Date;

    public getBusiness: Sequelize.BelongsToGetAssociationMixin<Businesses>;
    public setBusiness: Sequelize.BelongsToSetAssociationMixin<Businesses, Businesses['id']>;
    public createBusiness: Sequelize.BelongsToCreateAssociationMixin<IBusinessEntity>;

    public getCategory: Sequelize.BelongsToGetAssociationMixin<InternshipTypes>;
    public setCategory: Sequelize.BelongsToSetAssociationMixin<
        InternshipTypes,
        InternshipTypes['id']
    >;
    public createCategory: Sequelize.BelongsToCreateAssociationMixin<IInternshipTypeEntity>;

    public readonly business?: Businesses | Businesses['id'];
    public readonly category?: InternshipTypes | InternshipTypes['id'];
}

Internships.init(
    {
        id: {
            type: Sequelize.DataTypes.INTEGER.UNSIGNED,
            autoIncrement: true,
            primaryKey: true,
        },

        // Data
        subject: {
            type: new Sequelize.DataTypes.STRING(256),
            allowNull: false,
        },
        description: {
            type: new Sequelize.DataTypes.TEXT(),
            allowNull: true,
        },

        // Localisation
        country: {
            type: new Sequelize.DataTypes.STRING(128),
            allowNull: false,
        },
        city: {
            type: new Sequelize.DataTypes.STRING(128),
            allowNull: false,
        },
        postalCode: {
            type: new Sequelize.DataTypes.STRING(128),
            allowNull: false,
        },
        address: {
            type: new Sequelize.DataTypes.STRING(128),
            allowNull: false,
        },
        additional: {
            type: new Sequelize.DataTypes.STRING(),
            allowNull: true,
        },

        // State
        isInternshipAbroad: {
            type: Sequelize.DataTypes.BOOLEAN,
            allowNull: false,
            defaultValue: false,
        },
        isValidated: {
            type: Sequelize.DataTypes.BOOLEAN,
            allowNull: false,
            defaultValue: false,
        },
    },
    {
        tableName: 'internships',
        sequelize: database,
        defaultScope: {
            attributes: { exclude: ['businessId', 'categoryId'] },
        },
    },
);

export default Internships;
