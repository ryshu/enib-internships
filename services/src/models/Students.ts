import { Model, DataTypes } from 'sequelize';

import database from '../configs/instances/database';

class Students extends Model implements IStudentEntity {
    public id!: number; // Note that the `null assertion` `!` is required in strict mode.

    public firstName: string;
    public lastName: string;
    public email: string;
    public semester: string;

    // timestamps!
    public readonly createdAt!: Date;
    public readonly updatedAt!: Date;
}

Students.init(
    {
        id: {
            type: DataTypes.INTEGER.UNSIGNED,
            autoIncrement: true,
            primaryKey: true,
        },
        firstName: {
            type: new DataTypes.STRING(128),
            allowNull: false,
        },
        lastName: {
            type: new DataTypes.STRING(128),
            allowNull: false,
        },
        email: {
            type: new DataTypes.STRING(128),
            allowNull: false,
        },
        semester: {
            type: new DataTypes.STRING(3),
            allowNull: false,
        },
    },
    {
        tableName: 'students',
        sequelize: database,
    },
);

export default Students;
