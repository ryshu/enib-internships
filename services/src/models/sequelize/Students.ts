import * as Sequelize from 'sequelize';

import database from '../../configs/instances/database';

import { IInternshipEntity } from '../../declarations';

import Internships from './Internships';
class Students extends Sequelize.Model {
    public static associations: {
        student: Sequelize.Association<Students, Internships>;
    };

    public id!: number; // Note that the `null assertion` `!` is required in strict mode.

    public firstName: string;
    public lastName: string;
    public email: string;
    public semester: string;

    // timestamps!
    public readonly createdAt!: Date;
    public readonly updatedAt!: Date;

    // Internships
    public getInternships: Sequelize.HasManyGetAssociationsMixin<Internships>;
    public addInternship: Sequelize.HasManyAddAssociationMixin<Internships, Internships['id']>;
    public createInternship: Sequelize.HasManyCreateAssociationMixin<IInternshipEntity>;
    public hasInternship: Sequelize.HasManyHasAssociationMixin<Internships, Internships['id']>;
    public countInternships: Sequelize.HasManyCountAssociationsMixin;

    public readonly internships?: Internships[];
    public internshipCount: number;
}

Students.init(
    {
        id: {
            type: Sequelize.DataTypes.INTEGER.UNSIGNED,
            autoIncrement: true,
            primaryKey: true,
        },
        firstName: {
            type: new Sequelize.DataTypes.STRING(128),
            allowNull: false,
        },
        lastName: {
            type: new Sequelize.DataTypes.STRING(128),
            allowNull: false,
        },
        email: {
            type: new Sequelize.DataTypes.STRING(128),
            allowNull: false,
        },
        semester: {
            type: new Sequelize.DataTypes.STRING(3),
            allowNull: false,
        },
    },
    {
        tableName: 'students',
        sequelize: database,
    },
);

export default Students;
