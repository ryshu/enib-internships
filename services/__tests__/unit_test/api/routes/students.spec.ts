import request from 'supertest';

// Include app for super agent, not server
import app from '../../../../src/app';

// Include database connection
import dbSetup from '../../../../src/configs/setup/database';

// Import model for pre-operation before asserting API methods
import Students from '../../../../src/models/sequelize/Students';
import Internships from '../../../../src/models/sequelize/Internships';

import {
    defaultStudents,
    defaultInternships,
    defaultInternshipTypes,
} from '../../../../__mocks__/mockData';

jest.setTimeout(30000);

beforeAll((done) => {
    dbSetup.then(() => done()).catch((e) => done(e));
});

describe('GET /students', () => {
    beforeEach(async () => {
        // Remove all
        await Students.destroy({ where: {} });
    });

    it('NoStudent_204', async () => {
        const RESPONSE = await request(app).get(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/students`,
        );
        expect(RESPONSE.status).toBe(204);
    });

    it('BadRequest_400', async () => {
        const RESPONSE = await request(app).get(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/students?limit=Nan`,
        );
        expect(RESPONSE.status).toBe(400);
    });

    it('Students_200', async () => {
        const VALID_STUDENT = defaultStudents();

        await Students.create(VALID_STUDENT);
        const RESPONSE = await request(app).get(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/students`,
        );
        expect(RESPONSE.status).toBe(200);
        expect(Array.isArray(RESPONSE.body.data)).toBeTruthy();
        expect(RESPONSE.body).toMatchSnapshot({
            data: [
                {
                    createdAt: expect.any(String),
                    updatedAt: expect.any(String),
                },
            ],
        });
    });
});

describe('POST /students', () => {
    it('NoBody_400', async () => {
        const RESPONSE = await request(app).post(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/students`,
        );
        expect(RESPONSE.status).toBe(400);
    });

    it('ValidStudents_200', async () => {
        const VALID_STUDENT = defaultStudents();

        const RESPONSE = await request(app)
            .post(`/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/students`)
            .send(VALID_STUDENT);
        expect(RESPONSE.status).toBe(200);
        expect(RESPONSE.body).toMatchSnapshot({
            createdAt: expect.any(String),
            updatedAt: expect.any(String),
            id: expect.any(Number),
        });
    });
});

describe('GET /students/:id', () => {
    beforeEach(async () => {
        // Remove all
        await Students.destroy({ where: {} });
    });

    it('NoStudent_204', async () => {
        const RESPONSE = await request(app).get(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/students/10`,
        );
        expect(RESPONSE.status).toBe(204);
    });

    it('BadRequest_400', async () => {
        const RESPONSE = await request(app).get(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/students/{falseEncoding}`,
        );
        expect(RESPONSE.status).toBe(400);
    });

    it('Students_200', async () => {
        const VALID_STUDENT = defaultStudents();

        const CREATED = await Students.create(VALID_STUDENT);
        const RESPONSE = await request(app).get(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/students/${CREATED.id}`,
        );
        expect(RESPONSE.status).toBe(200);
        expect(RESPONSE.body).toMatchSnapshot({
            createdAt: expect.any(String),
            updatedAt: expect.any(String),
            id: expect.any(Number),
        });
    });
});

describe('PUT /students/:id', () => {
    beforeEach(async () => {
        // Remove all
        await Students.destroy({ where: {} });
    });

    it('NoStudent_204', async () => {
        const RESPONSE = await request(app).put(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/students/10`,
        );
        expect(RESPONSE.status).toBe(204);
    });

    it('BadRequest_400', async () => {
        const RESPONSE = await request(app).put(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/students/{falseEncoding}`,
        );
        expect(RESPONSE.status).toBe(400);
    });

    it('Students_200_UpdateAllData', async () => {
        const VALID_STUDENT = defaultStudents();

        const CREATED = await Students.create(VALID_STUDENT);

        // Change some data on VALID_STUDENT
        VALID_STUDENT.lastName = 'Done';

        const RESPONSE = await request(app)
            .put(`/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/students/${CREATED.id}`)
            .send(VALID_STUDENT);
        expect(RESPONSE.status).toBe(200);
        expect(RESPONSE.body).toMatchSnapshot({
            createdAt: expect.any(String),
            updatedAt: expect.any(String),
            id: expect.any(Number),
        });
    });

    it('Students_200_NotAnyDataUpdated', async () => {
        const VALID_STUDENT = defaultStudents();

        const CREATED = await Students.create(VALID_STUDENT);

        const RESPONSE = await request(app)
            .put(`/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/students/${CREATED.id}`)
            .send({});
        expect(RESPONSE.status).toBe(200);
        expect(RESPONSE.body).toMatchSnapshot({
            createdAt: expect.any(String),
            updatedAt: expect.any(String),
            id: expect.any(Number),
        });
    });
});

describe('DELETE /students/:id', () => {
    beforeEach(async () => {
        // Remove all
        await Students.destroy({ where: {} });
    });

    it('NoStudent_200', async () => {
        const RESPONSE = await request(app).delete(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/students/10`,
        );
        expect(RESPONSE.status).toBe(200);
    });

    it('BadRequest_400', async () => {
        const RESPONSE = await request(app).delete(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/students/{falseEncoding}`,
        );
        expect(RESPONSE.status).toBe(400);
    });

    it('Students_200', async () => {
        const VALID_STUDENT = defaultStudents();

        const CREATED = await Students.create(VALID_STUDENT);

        const RESPONSE = await request(app).delete(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/students/${CREATED.id}`,
        );
        expect(RESPONSE.status).toBe(200);
    });
});

describe('GET /students/:id/internships', () => {
    beforeEach(async () => {
        // Remove all
        await Students.destroy({ where: {} });
    });

    it('NoStudent_204', async () => {
        const RESPONSE = await request(app).get(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/students/10/internships`,
        );
        expect(RESPONSE.status).toBe(204);
    });

    it('BadRequest_400', async () => {
        const RESPONSE = await request(app).get(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/students/{falseEncoding}/internships`,
        );
        expect(RESPONSE.status).toBe(400);
    });

    it('Students_204_NoLinkedData', async () => {
        const VALID_STUDENT = defaultStudents();

        const CREATED = await Students.create(VALID_STUDENT);
        const RESPONSE = await request(app).get(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/students/${CREATED.id}/internships`,
        );
        expect(RESPONSE.status).toBe(204);
    });

    it('Students_200_WithLinkedData', async () => {
        const VALID_STUDENT = defaultStudents();
        const VALID_INTERNSHIP = defaultInternships();
        const VALID_INTERNSHIP_TYPE = defaultInternshipTypes();

        VALID_INTERNSHIP.category = VALID_INTERNSHIP_TYPE;
        VALID_STUDENT.internships = [VALID_INTERNSHIP];

        const CREATED_STUDENT = await Students.create(VALID_STUDENT, {
            include: [
                {
                    association: Students.associations.internships,
                    include: [{ association: Internships.associations.category }],
                },
            ],
        });

        const RESPONSE = await request(app).get(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/students/${CREATED_STUDENT.id}/internships`,
        );
        expect(RESPONSE.status).toBe(200);
        expect(RESPONSE.body).toHaveLength(1);
    });
});

describe('POST /students/:id/internships/:internship_id/link', () => {
    beforeEach(async () => {
        // Remove all
        await Students.destroy({ where: {} });
    });

    it('NoStudent_204', async () => {
        const RESPONSE = await request(app).post(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/students/10/internships/20/link`,
        );
        expect(RESPONSE.status).toBe(204);
    });

    it('BadRequest_400_WrongID', async () => {
        const RESPONSE = await request(app).post(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/students/{falseEncoding}/internships/10/link`,
        );
        expect(RESPONSE.status).toBe(400);
    });

    it('BadRequest_400_WrongInternshipID', async () => {
        const RESPONSE = await request(app).post(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/students/10/internships/{falseEncoding}/link`,
        );
        expect(RESPONSE.status).toBe(400);
    });

    it('Students_200_NoInternship', async () => {
        // In this case, we check if link a existing Bussiness and an unexisting internships work
        const VALID_STUDENT = defaultStudents();

        const CREATED = await Students.create(VALID_STUDENT);
        const RESPONSE = await request(app).post(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/students/${CREATED.id}/internships/20/link`,
        );
        expect(RESPONSE.status).toBe(204);
    });

    it('Students_200_WithInternship', async () => {
        const VALID_STUDENT = defaultStudents();
        const VALID_INTERNSHIP = defaultInternships();

        let CREATED_STUDENT = await Students.create(VALID_STUDENT);
        const CREATED_INTERNSHIP = await Internships.create(VALID_INTERNSHIP);

        const RESPONSE = await request(app).post(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/students/${CREATED_STUDENT.id}/internships/${CREATED_INTERNSHIP.id}/link`,
        );

        // Should answer 200
        expect(RESPONSE.status).toBe(200);

        // check if student and internship are linked
        CREATED_STUDENT = await Students.findByPk(CREATED_STUDENT.id);

        const internships = await CREATED_STUDENT.getInternships();
        expect(internships).toHaveLength(1);
        expect(internships[0].id).toBe(CREATED_INTERNSHIP.id);
    });
});
