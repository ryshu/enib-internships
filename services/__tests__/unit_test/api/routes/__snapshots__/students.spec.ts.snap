// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET /students Students_200 1`] = `
Object {
  "data": Array [
    Object {
      "createdAt": Any<String>,
      "email": "john.doe@enib.fr",
      "firstName": "John",
      "id": 1,
      "lastName": "Doe",
      "semester": "S10",
      "updatedAt": Any<String>,
    },
  ],
  "length": 1,
  "max": 1,
  "page": 1,
}
`;

exports[`GET /students/:id Students_200 1`] = `
Object {
  "createdAt": Any<String>,
  "email": "john.doe@enib.fr",
  "firstName": "John",
  "id": Any<Number>,
  "lastName": "Doe",
  "semester": "S10",
  "updatedAt": Any<String>,
}
`;

exports[`POST /students ValidStudents_200 1`] = `
Object {
  "createdAt": Any<String>,
  "email": "john.doe@enib.fr",
  "firstName": "John",
  "id": Any<Number>,
  "lastName": "Doe",
  "semester": "S10",
  "updatedAt": Any<String>,
}
`;

exports[`PUT /students/:id Students_200_NotAnyDataUpdated 1`] = `
Object {
  "createdAt": Any<String>,
  "email": "john.doe@enib.fr",
  "firstName": "John",
  "id": Any<Number>,
  "lastName": "Doe",
  "semester": "S10",
  "updatedAt": Any<String>,
}
`;

exports[`PUT /students/:id Students_200_UpdateAllData 1`] = `
Object {
  "createdAt": Any<String>,
  "email": "john.doe@enib.fr",
  "firstName": "John",
  "id": Any<Number>,
  "lastName": "Done",
  "semester": "S10",
  "updatedAt": Any<String>,
}
`;
